File: app/core/export.py
import csv
import os

import numpy as np

from app.core.atomic_io import atomic_open, atomic_save_npy


def _clean_coords(coords_fw, active_mask):
    coords_fw = np.asarray(coords_fw)
    active_mask = np.asarray(active_mask, dtype=bool)
    if coords_fw.ndim != 3 or coords_fw.shape[2] != 2:
        raise ValueError("Coordinates must have shape (frames, points, 2)")
    if active_mask.shape != (coords_fw.shape[1],):
        raise ValueError("Active mask length does not match the coordinate point axis")
    if not np.isfinite(coords_fw[:, active_mask, :]).all():
        raise ValueError("Selected export coordinates contain NaN or infinity")
    return coords_fw[:, active_mask, :].astype(np.float32)


def export(
    coords_fw: np.ndarray,
    active_mask: np.ndarray,
    reference_index: int,
    last_index: int,
    out_dir: str,
    filename: str = "coords.npy",
):
    """Write cleaned forward coordinates and the sequence range.

    - coords.npy : forward coords for kept points, shape (N_frames, N_points_kept, 2)
    - sequence.txt : "<reference_index> <last_index>" (global indices)

    Returns (coords_path, sequence_path, coords_shape).
    """
    if not filename.endswith(".npy"):
        filename += ".npy"

    coords = _clean_coords(coords_fw, active_mask)
    coords_path = os.path.join(out_dir, filename)
    if reference_index < 0 or last_index - reference_index + 1 != coords.shape[0]:
        raise ValueError("Export frame range does not match coordinates")
    # One atomic file is authoritative; loose arrays/text below are legacy convenience views.
    with atomic_open(os.path.splitext(coords_path)[0] + ".bundle.npz", "wb") as handle:
        np.savez(handle, coords=coords, point_ids=np.flatnonzero(active_mask),
                 frame_indices=np.arange(reference_index, last_index + 1))
    atomic_save_npy(coords_path, coords)

    sequence_path = os.path.join(out_dir, "sequence.txt" if filename == "coords.npy" else os.path.splitext(filename)[0] + ".sequence.txt")
    with atomic_open(sequence_path) as f:
        f.write(f"{reference_index} {last_index}\n")

    return coords_path, sequence_path, coords.shape


def export_csv(
    coords_fw: np.ndarray,
    active_mask: np.ndarray,
    frame_names,
    out_dir: str,
    filename: str = "coords.csv",
):
    """Write cleaned forward coords as CSV: one row per frame, columns interleaved x/y.

    Header:  filename,p1x,p1y,p2x,p2y,...
    Row i :  <frame_names[i]>, x0, y0, x1, y1, ...   for the kept points only.

    Returns (csv_path, coords_shape).
    """
    if not filename.endswith(".csv"):
        filename += ".csv"

    coords = _clean_coords(coords_fw, active_mask)  # (N, P_kept, 2)
    n_frames, n_pts, _ = coords.shape
    if len(frame_names) != n_frames:
        raise ValueError("Frame-name count does not match the coordinate frame axis")
    flat = coords.reshape(n_frames, n_pts * 2)  # p0x,p0y,p1x,p1y,...

    header = ["filename"]
    for p in range(n_pts):
        header += [f"p{p + 1}x", f"p{p + 1}y"]  # 1-based labels

    csv_path = os.path.join(out_dir, filename)
    with atomic_open(csv_path, newline="") as f:
        writer = csv.writer(f)
        writer.writerow(header)
        for i in range(n_frames):
            # Fixed 6-decimal format (matches the affine_zones exporter); repr() would emit
            # noisy full-precision float64 expansions of the float32 values.
            writer.writerow([frame_names[i], *(f"{v:.6f}" for v in flat[i])])

    return csv_path, coords.shape

File: rust/crates/core/src/export.rs
//! Export cleaned forward coordinates to `.npy` (+ `sequence.txt`) and `.csv`.
//! Ports `app/core/export.py`. (CSV uses `\n` line endings rather than Python csv's `\r\n`.)

use ndarray::{Array3, Axis};
use ndarray_npy::write_npy;
use std::io::Write;
use std::path::{Path, PathBuf};

fn kept_indices(active_mask: &[bool]) -> Vec<usize> {
    active_mask
        .iter()
        .enumerate()
        .filter(|(_, &k)| k)
        .map(|(i, _)| i)
        .collect()
}

/// Forward coords for kept points: `(N, P_kept, 2)` f32.
fn kept_coords(coords_fw: &Array3<f32>, active_mask: &[bool]) -> Array3<f32> {
    coords_fw.select(Axis(1), &kept_indices(active_mask))
}

fn npy_err(e: ndarray_npy::WriteNpyError) -> std::io::Error {
    std::io::Error::new(std::io::ErrorKind::Other, e.to_string())
}

pub struct NpyExport {
    pub coords_path: PathBuf,
    pub sequence_path: PathBuf,
    pub shape: (usize, usize, usize),
}

/// Write `coords.npy` (kept forward coords) + `sequence.txt` (`"<ref> <last>"`, global indices).
pub fn export(
    coords_fw: &Array3<f32>,
    active_mask: &[bool],
    reference_index: usize,
    last_index: usize,
    out_dir: &Path,
    filename: &str,
) -> std::io::Result<NpyExport> {
    let mut fname = filename.to_string();
    if !fname.ends_with(".npy") {
        fname.push_str(".npy");
    }
    let coords = kept_coords(coords_fw, active_mask);
    let coords_path = out_dir.join(&fname);
    write_npy(&coords_path, &coords).map_err(npy_err)?;

    let sequence_path = out_dir.join("sequence.txt");
    std::fs::write(&sequence_path, format!("{reference_index} {last_index}\n"))?;

    let s = coords.shape();
    Ok(NpyExport {
        coords_path,
        sequence_path,
        shape: (s[0], s[1], s[2]),
    })
}

/// Write kept forward coords as CSV: one row per frame, columns `filename,p1x,p1y,p2x,p2y,...`.
pub fn export_csv(
    coords_fw: &Array3<f32>,
    active_mask: &[bool],
    frame_names: &[String],
    out_dir: &Path,
    filename: &str,
) -> std::io::Result<(PathBuf, (usize, usize, usize))> {
    let mut fname = filename.to_string();
    if !fname.ends_with(".csv") {
        fname.push_str(".csv");
    }
    let coords = kept_coords(coords_fw, active_mask); // (N, P_kept, 2)
    let n = coords.shape()[0];
    let p = coords.shape()[1];

    let csv_path = out_dir.join(&fname);
    let mut f = std::fs::File::create(&csv_path)?;

    let mut header = String::from("filename");
    for pi in 0..p {
        header.push_str(&format!(",p{}x,p{}y", pi + 1, pi + 1)); // 1-based labels
    }
    writeln!(f, "{header}")?;

    for i in 0..n {
        let mut row = frame_names.get(i).cloned().unwrap_or_default();
        for j in 0..p {
            // Fixed 6-decimal format (matches the Python exporter).
            row.push_str(&format!(",{:.6},{:.6}", coords[[i, j, 0]], coords[[i, j, 1]]));
        }
        writeln!(f, "{row}")?;
    }
    Ok((csv_path, (n, p, 2)))
}

#[cfg(test)]
mod tests {
    use super::*;
    use ndarray::Array;

    #[test]
    fn export_subsets_kept_points_and_roundtrips() {
        // 3 frames, 4 points; keep points 0 and 2.
        let coords = Array::from_shape_fn((3, 4, 2), |(t, j, c)| (t * 100 + j * 10 + c) as f32);
        let mask = vec![true, false, true, false];
        let dir = std::env::temp_dir().join(format!("ecm_export_{}", std::process::id()));
        std::fs::create_dir_all(&dir).unwrap();

        let out = export(&coords, &mask, 5, 16, &dir, "coords").unwrap();
        assert_eq!(out.shape, (3, 2, 2));
        assert_eq!(
            std::fs::read_to_string(&out.sequence_path).unwrap(),
            "5 16\n"
        );

        // Round-trip the npy and confirm it holds exactly the kept columns.
        let back: Array3<f32> = ndarray_npy::read_npy(&out.coords_path).unwrap();
        assert_eq!(back.shape(), &[3, 2, 2]);
        for t in 0..3 {
            for (out_j, src_j) in [0usize, 2usize].into_iter().enumerate() {
                assert_eq!(back[[t, out_j, 0]], coords[[t, src_j, 0]]);
                assert_eq!(back[[t, out_j, 1]], coords[[t, src_j, 1]]);
            }
        }

        let names: Vec<String> = (0..3).map(|i| format!("img_{i}.png")).collect();
        let (csv_path, shape) = export_csv(&coords, &mask, &names, &dir, "coords").unwrap();
        assert_eq!(shape, (3, 2, 2));
        let text = std::fs::read_to_string(&csv_path).unwrap();
        let lines: Vec<&str> = text.lines().collect();
        assert_eq!(lines[0], "filename,p1x,p1y,p2x,p2y");
        // frame 0: point0=(0,1), point2=(20,21)
        assert_eq!(lines[1], "img_0.png,0.000000,1.000000,20.000000,21.000000");

        std::fs::remove_dir_all(&dir).ok();
    }
}
